File: blind75/28_course_schedule/course_schedule.py
from typing import List
from collections import defaultdict, deque


def can_finish(num_courses: int, prerequisites: List[List[int]]) -> bool:
    """
    Determines if all courses can be completed using DFS cycle detection.
    Time: O(V + E), Space: O(V + E)

    Args:
        num_courses: Total number of courses
        prerequisites: List of [course, prerequisite] pairs

    Returns:
        True if all courses can be completed, False otherwise
    """
    # Build adjacency list
    graph = defaultdict(list)
    for course, prereq in prerequisites:
        graph[prereq].append(course)

    # 0 = unvisited, 1 = visiting, 2 = visited
    state = [0] * num_courses

    def has_cycle(course: int) -> bool:
        if state[course] == 1:
            # Currently visiting - found a cycle
            return True
        if state[course] == 2:
            # Already visited
            return False

        # Mark as visiting
        state[course] = 1

        # Check all neighbors
        for neighbor in graph[course]:
            if has_cycle(neighbor):
                return True

        # Mark as visited
        state[course] = 2
        return False

    # Check each course
    for i in range(num_courses):
        if has_cycle(i):
            return False

    return True


def can_finish_bfs(num_courses: int, prerequisites: List[List[int]]) -> bool:
    """
    Determines if all courses can be completed using Kahn's algorithm (BFS).
    Time: O(V + E), Space: O(V + E)

    Args:
        num_courses: Total number of courses
        prerequisites: List of [course, prerequisite] pairs

    Returns:
        True if all courses can be completed, False otherwise
    """
    # Build adjacency list and in-degree array
    graph = defaultdict(list)
    in_degree = [0] * num_courses

    for course, prereq in prerequisites:
        graph[prereq].append(course)
        in_degree[course] += 1

    # Find all courses with no prerequisites
    queue = deque([i for i in range(num_courses) if in_degree[i] == 0])

    courses_completed = 0

    # Process courses
    while queue:
        current = queue.popleft()
        courses_completed += 1

        # Reduce in-degree for all dependent courses
        for neighbor in graph[current]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return courses_completed == num_courses

File: blind75/28_course_schedule/course_schedule.go
package course_schedule

// CanFinish determines if all courses can be completed using DFS cycle detection
// Time: O(V + E), Space: O(V + E)
func CanFinish(numCourses int, prerequisites [][]int) bool {
	// Build adjacency list
	graph := make([][]int, numCourses)
	for _, prereq := range prerequisites {
		course, prerequisite := prereq[0], prereq[1]
		graph[prerequisite] = append(graph[prerequisite], course)
	}

	// 0 = unvisited, 1 = visiting, 2 = visited
	state := make([]int, numCourses)

	var hasCycle func(course int) bool
	hasCycle = func(course int) bool {
		if state[course] == 1 {
			// Currently visiting - found a cycle
			return true
		}
		if state[course] == 2 {
			// Already visited
			return false
		}

		// Mark as visiting
		state[course] = 1

		// Check all neighbors
		for _, neighbor := range graph[course] {
			if hasCycle(neighbor) {
				return true
			}
		}

		// Mark as visited
		state[course] = 2
		return false
	}

	// Check each course
	for i := 0; i < numCourses; i++ {
		if hasCycle(i) {
			return false
		}
	}

	return true
}

// CanFinishBFS determines if all courses can be completed using Kahn's algorithm (BFS)
// Time: O(V + E), Space: O(V + E)
func CanFinishBFS(numCourses int, prerequisites [][]int) bool {
	// Build adjacency list and in-degree array
	graph := make([][]int, numCourses)
	inDegree := make([]int, numCourses)

	for _, prereq := range prerequisites {
		course, prerequisite := prereq[0], prereq[1]
		graph[prerequisite] = append(graph[prerequisite], course)
		inDegree[course]++
	}

	// Find all courses with no prerequisites
	queue := []int{}
	for i := 0; i < numCourses; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	coursesCompleted := 0

	// Process courses
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		coursesCompleted++

		// Reduce in-degree for all dependent courses
		for _, neighbor := range graph[current] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	return coursesCompleted == numCourses
}
